internal/cri/server/container_image_mount.go

```diff
@@ -0,0 +1,213 @@
+		return fmt.Errorf("readonly must be true while mount image: %+v", extraMount)
+	}
+
+	ref := imageSpec.GetImage()
+	if ref == "" {
+		return fmt.Errorf("image not specified in: %+v", imageSpec)
+	}
+	image, err := c.LocalResolve(ref)
+	if err != nil {
+		return fmt.Errorf("failed to resolve image %q: %w", ref, err)
+	}
+	containerdImage, err := c.toContainerdImage(ctx, image)
+	if err != nil {
+		return fmt.Errorf("failed to get image from containerd %q: %w", image.ID, err)
+	}
+
+	// This is a digest of the manifest
+	imageID := containerdImage.Target().Digest.Hex()
+
+	target := c.getImageVolumeHostPath(sandboxID, imageID)
+
+	// Already mounted in another container on the same pod
+	if stat, err := os.Stat(target); err == nil && stat.IsDir() {
+		extraMount.HostPath = target
+		return nil
+	}
+
+	ctx, done, err := c.client.WithLease(ctx,
+		leases.WithID(target),
+		leases.WithLabel(defaults.DefaultSnapshotterNSLabel, snapshotter),
+		leases.WithLabel(labelGCSnapRef+snapshotter, target),
+	)
+	if err != nil && !errdefs.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create lease: %w", err)
+	}
+	defer func() {
+		if retErr != nil {
+			_ = done(ctx) // call lease done func.
+		}
+	}()
+
+	img, err := c.client.ImageService().Get(ctx, ref)
+	if err != nil {
+		return fmt.Errorf("failed to get image volume ref %q: %w", ref, err)
+	}
+
+	i := containerd.NewImageWithPlatform(c.client, img, platforms.Only(platform))
+	if err := i.Unpack(ctx, snapshotter); err != nil {
+		return fmt.Errorf("failed to unpack image volume: %w", err)
+	}
+
+	diffIDs, err := i.RootFS(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get diff IDs for image volume %q: %w", ref, err)
+	}
+	chainID := identity.ChainID(diffIDs).String()
+
+	s := c.client.SnapshotService(snapshotter)
+	mounts, err := s.Prepare(ctx, target, chainID)
+	if err != nil {
+		return fmt.Errorf("failed to prepare for image volume %q: %w", ref, err)
+	}
+	defer func() {
+		if retErr != nil {
+			_ = s.Remove(ctx, target)
+		}
+	}()
+
+	err = os.MkdirAll(target, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create directory to image volume target path %q: %w", target, err)
+	}
+
+	if err := mount.All(mounts, target); err != nil {
+		return fmt.Errorf("failed to mount image volume component %q: %w", target, err)
+	}
+
+	extraMount.HostPath = target
+	return nil
+}
+
+func (c *criService) cleanupImageMounts(
+	ctx context.Context,
+	sandboxID string,
+) (retErr error) {
+	// Some checks to avoid affecting old pods.
+	ociRuntime, err := c.getPodSandboxRuntime(sandboxID)
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to get sandbox runtime handler %q", sandboxID)
+		return nil
+	}
+	snapshotter := c.RuntimeSnapshotter(ctx, ociRuntime)
+	if snapshotter == "" {
+		return nil
+	}
+	s := c.client.SnapshotService(snapshotter)
+	if s == nil {
+		return nil
+	}
+	targetBase := c.getImageVolumeBaseDir(sandboxID)
+	dir, err := os.ReadDir(targetBase)
```

Comment:
https://github.com/containerd/containerd/actions/runs/11439133722/job/31822309174

``` log
failed to delete lease "C:\\ProgramData\\containerd\\state-test\\io.containerd.grpc.v1.cri\\image-volumes\\22c62ef43d85f06ffdebbf8dc1e95b1cc7e60f7b9113ad0a8be44584755034fa\\184914~1": not found
```


I'm getting short file names here on Windows. Is there a better way?

Reply:
cc @jsturtevant

Reply:
gosh I missed the notification, Looks like this changed a bit since this happened.  Are you still experience trouble?

Reply:
Not yet.

I tried to get long files based on short files, but it only works on Windows 2022, not Windows 2019.

After Containerd 2.0, there are new errors blocking the cleanup mount dir on Windows 2022.

Reply:
The ws2019 passed, it looks like an error in clean up the job.  WS2022 looks like a flake in the restart test but otherwise looks like tests are passing, could we trigger again to see if we get different results?

Reply:
I reproduced the error on WS2022, It looks like the there is a timing issue when removing the ImageMounts.  I believe the snapshotter is holding onto the file handle....

Reply:
There was an `Open()` call on the dir but missing the `Close()`, once that was added in the tests pass on WS2022.

Reply:
this might fix ws2019 as well. If it doesn't I'll have to get an environment set up.  

@kiashok What's the plan for support for WS2019 in containerd?  We Containerd 2.0 support WS2019 since it is EOL now?